[Experimental] Scaffolding Taxonomy Filter block

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR is the first step implementing the Taxonomy Filter block by scaffolding and registering the block. Beside block registration, this PR:
- registers block variations for each taxonomy.
- registers all necessary block inspector settings and ensure the editor preview work with them.
- extract the display style switcher into a shared component for refactoring other filter block later.

Closes wooplug-4764, closes #59128 

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

<img width="1025" alt="image" src="https://github.com/user-attachments/assets/ab69492a-b5c9-412d-8dd6-d62f8d7080fa" />

<img width="1025" alt="image" src="https://github.com/user-attachments/assets/9006dd52-7968-48ad-8b61-17423f2cc7b8" />


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable `experimental-blocks` flag using WC Beta Tester.
2. Add Product Filters block to the Product Catalog template.
3. Focus on the Product Filters block.
4. Open the block inserter, search for `brand filter`.
5. See the Brand block with specific description for brand.
6. Insert that block to the Product Filters block.
7. Try changing the taxonomy, display style and other settings, see the editor preview reacts accordingly.
8. Save and reload the editor, see the updated settings preserves.
9. See only the heading is rendered on the frontend.

<!-- End testing instructions -->

dinhtungdu committed:
feat: scaffold

dinhtungdu committed:
refactor: update filter heading utility

dinhtungdu committed:
refactor: use select for shared component

dinhtungdu committed:
fix: put the block under experimental flag

dinhtungdu committed:
chore: copy

dinhtungdu committed:
chore: changelog

dinhtungdu committed:
fix: setting type

dinhtungdu committed:
chore: phpcs

kmanijak commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/inner-blocks/taxonomy-filter/edit.tsx:
```diff
@@ -0,0 +1,123 @@
+		}
+	);
+
+	const isLoading = isOptionsLoading;
```
Nitpick, but assigning a constant to essentially rename it feels super redundant 😅

dinhtungdu replied:
Haha, it is. We will have more loading status combined to `isLoading` but we can do that later, I will fix this before merging.

kmanijak commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/inner-blocks/taxonomy-filter/inspector.tsx:
```diff
@@ -0,0 +1,111 @@
+import { updateFilterHeading } from '../../utils/update-filter-heading';
+import { getTaxonomyLabel } from './utils';
+
+export const Inspector = ( {
```
Personally, I find name `Inspector` confusing. If possible, I'd rather avoid naming `InspectorControls` as `Inspector`. I know it comes from conflict with `InspectorControls` from `'@wordpress/block-editor'`. But if you don't mind, what about changing this export to default export:

```suggestion
export default function ( {
```

and import it as `InspectorControls` or give it more explicit name:

```suggestion
export const TaxonomyFilterInspectorControls = ( {
```

dinhtungdu replied:
Will do :D

kmanijak commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/inner-blocks/taxonomy-filter/inspector.tsx:
```diff
@@ -0,0 +1,111 @@
+
+	return (
+		<InspectorControls>
+			<PanelBody title={ __( 'Taxonomy', 'woocommerce' ) }>
```
We should be switching from using `PanelBody` to `ToolsPanel` (https://github.com/woocommerce/woocommerce/issues/59464)

dinhtungdu replied:
TIL, thanks!

kmanijak commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/inner-blocks/taxonomy-filter/utils.ts:
```diff
@@ -0,0 +1,21 @@
+
+export function getTaxonomyLabel( taxonomy: string ) {
+	const match = taxonomies.find( ( item ) => item.name === taxonomy );
+	if ( match ) return match.label;
```
I'm surprised linter passed, I was pretty sure we're forcing brackets around `if` statement body! 😆 

```suggestion
	if ( match ) {
	    return match.label;
	}
```

kmanijak commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/inner-blocks/taxonomy-filter/index.tsx:
```diff
@@ -0,0 +1,52 @@
+	variations: taxonomies.map( ( item, index ) => {
+		return {
+			name: `product-filter-taxonomy-${ item.name }`,
+			title: item.label,
```
I'd add `filter` suffix ASAP as inserter becomes confusing straight away 😅 

```suggestion
			title: `${ item.label } Filter`,
```

<img width="207" alt="image" src="https://github.com/user-attachments/assets/ffa20bdb-0a2b-4feb-86b7-8e2a62eede2c" />

<img width="222" alt="image" src="https://github.com/user-attachments/assets/d48c9b02-011a-4015-9562-5d5522cfe386" />

dinhtungdu replied:
Haha, yeah, I will do it. Attribute filter variations also need the same.

kmanijak reviewed (approved):
I left couple of suggestions and findings but overall nice scaffolding work! 

I tested the PR and checked blocks (variations) are nicely listed in inserter:

<img width="431" alt="image" src="https://github.com/user-attachments/assets/389eddce-0872-41ae-9cff-59e5295f1006" />

They can be added and display the default data for now. One filter can be transformed into another using select control and on editor there's heading preserved as expected!

Given this is behind experimental flag and comments are rather minor I decided to pre-approve however, feel free to ping me in case you want another round of review once you apply some changes 🙌 Good work!

dinhtungdu committed:
refactor: remove redundant isLoading variable assignment

Address review feedback by removing unnecessary variable assignment
and using isOptionsLoading directly.

dinhtungdu committed:
refactor: rename Inspector to TaxonomyFilterInspectorControls

Address review feedback by giving the component a more descriptive name
to avoid confusion with the InspectorControls from @wordpress/block-editor.

dinhtungdu committed:
refactor: replace PanelBody with ToolsPanel in inspector

Address review feedback by switching from PanelBody to ToolsPanel and
ToolsPanelItem components for better consistency with modern patterns.
Each control now properly implements hasValue and onDeselect methods.

dinhtungdu committed:
style: add brackets around if statement body

Address review feedback by adding brackets around the if statement body
to follow coding standards and improve readability.

dinhtungdu committed:
feat: add 'Filter' suffix to variation titles

Address review feedback by adding 'Filter' suffix to block variation titles
to avoid confusion in the block inserter and make the purpose clearer.

dinhtungdu committed:
test: taxonomy filter integration test

- Add act wrapper for async fireEvent calls
- Fix block selector from 'Product Categories Filter' to 'Taxonomy'
- Add checkbox-list component import
- Remove preview mode test that was causing issues
- Improve test reliability and async handling

dinhtungdu committed:
chore: rename block to add Filter suffix

dinhtungdu committed:
test: update integration to check for actual change in the editor preview

dinhtungdu committed:
test(taxonomy-filter): extract dropdown menu interaction pattern and use regex for count assertions

dinhtungdu committed:
fix: default display style

dinhtungdu committed:
feat(product-filters): add reset functionality for display style blocks

- Add resetDisplayStyleBlock function to handle proper block reset behavior
- Refactor taxonomy filter inspector to use metadata defaults instead of hardcoded values
- Move static taxonomies data outside component for better performance
- Add consistent reset behavior for both resetAll and onDeselect callbacks

dinhtungdu committed:
chore: address coderabbit review

dinhtungdu committed:
chore: add taxonomy filter to allowed widget area block
